### PR TITLE
Created CompatibilityMetric class, added register_alias decorator

### PR DESCRIPTION
**Pull request by jneeven**
CompatibilityMetric will be the parent class of all our custom metrics. I added a register_alias decorator and registered the FlipRatio under flip_ratio, such that we can find it with tf.keras.metrics.get() in the future.

### CODE REVIEW

File: larq/metrics.py [hunk 1]
```diff
@@ -62,7 +63,54 @@ def get_training_metrics():
...
+class _CompatibilityMetric(Metric):
+    r"""Metric with support for both 1.13 and 1.14+
+    """
+
+    def get_config(self):
+        """Returns the serializable config of the metric."""
+        return {**super().get_config(), "name": self.name, "dtype": self.dtype}
+
+    def reset_states(self):
+        return K.batch_set_value([(v, 0) for v in self.variables])
```
Comment (lgeiger) [inline larq/metrics.py]: Since we only use it here, I'd prefer:
```suggestion
        return tf.keras.backend.batch_set_value([(v, 0) for v in self.variables])
```
But that's just personal preference.

File: larq/metrics.py [hunk 2]
```diff
@@ -62,7 +63,54 @@ def get_training_metrics():
     return _GLOBAL_TRAINING_METRICS
 
 
-class FlipRatio(Metric):
+class _CompatibilityMetric(Metric):
```
Comment (lgeiger) [inline larq/metrics.py]: Thinking about this again what do you think about `LarqMetric`?

File: larq/metrics.py [hunk 3]
```diff
@@ -62,7 +63,54 @@ def get_training_metrics():
...
+    r"""Metric with support for both 1.13 and 1.14+
+    """
```
Comment (lgeiger) [inline larq/metrics.py]:
```suggestion
```

File: larq/metrics.py [hunk 4]
```diff
@@ -62,7 +63,54 @@ def get_training_metrics():
     return _GLOBAL_TRAINING_METRICS
 
 
-class FlipRatio(Metric):
+class _CompatibilityMetric(Metric):
+    r"""Metric with support for both 1.13 and 1.14+
```
Comment (lgeiger) [inline larq/metrics.py]:
```suggestion
    """Metric with support for both 1.13 and 1.14+"""
```

File: larq/metrics.py [hunk 5]
```diff
@@ -62,7 +63,54 @@ def get_training_metrics():
...
+    """
+
+    def get_config(self):
```
Comment (lgeiger) [inline larq/metrics.py]: I don't think we need to overwrite this since this is the default anyway:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L199-L201

File: larq/metrics.py [hunk 6]
```diff
@@ -62,7 +63,54 @@ def get_training_metrics():
...
+                dtype=self._dtype if dtype is None else dtype,
+                trainable=False,
+                initializer=initializer,
+                collections=[],
+                synchronization=synchronization,
+                aggregation=aggregation,
+            )
+
+
+@utils.register_alias("flip_ratio")
```
Comment (lgeiger) [inline larq/metrics.py]: Nice!

File: larq/metrics.py [hunk 7]
```diff
@@ -87,18 +135,24 @@ def __init__(
         super().__init__(name=name, dtype=dtype)
         self.values_dtype = tf.as_dtype(values_dtype)
         self.values_shape = tf.TensorShape(values_shape).as_list()
+        self.is_weight_metric = True
```
Comment (lgeiger) [inline larq/metrics.py]: This isn't used anywhere. Or am I missing something?
```suggestion
```

File: larq/metrics.py [hunk 8]
```diff
@@ -87,18 +135,24 @@ def __init__(
...
         with tf.init_scope():
             self._previous_values = self.add_weight(
                 "previous_values",
                 shape=values_shape,
                 dtype=self.values_dtype,
                 initializer=tf.keras.initializers.zeros,
+                aggregation=tf.VariableAggregation.SUM,
```
Comment (lgeiger) [inline larq/metrics.py]: I am not sure if this is correct. Could you elaborate on this change?

File: larq/metrics.py (hunk 7 above)
Comment (jneeven) [reply larq/metrics.py]: Not in this PR, but I need it for the MetricLayer. Didn't want to push them all at the same time since the changes might be a bit overwhelming then

File: larq/metrics.py (hunk 8 above)
Comment (jneeven) [reply larq/metrics.py]: You already had it on SUM before, it was the default value in the add_weight function. However, SUM may not always be the desired aggregation method, so I changed the default to None (which will throw an error if you run it on multi-gpu, rather than computing incorrect values) and explicitly set it back to SUM here. 
That being said, I think it is actually incorrect here; I don't think FlipRatio can support multi-GPU at all. In that case, I should set it to None. What do you think?

File: larq/metrics.py (hunk 2 above)
Comment (jneeven) [reply larq/metrics.py]: Sure, at this point there is nothing that really makes it larq-specific, but since all Larq metrics will subclass it, why not

File: larq/metrics.py (hunk 8 above)
Comment (jneeven) [reply larq/metrics.py]: Alternatively, we could set all of these to ONLY_FIRST_REPLICA, which will use only one of the results in the multi-gpu case and ignore the others.

Comment (jneeven) [reply larq/metrics.py]: I changed it to ONLY_FIRST_REPLICA in the latest commit. Since the FlipRatio uses weights instead of activations, I assume that even across multiple GPU's, either the metric would only be computed once, or all replicas would compute the same value.

File: larq/metrics.py [hunk 9]
```diff
@@ -62,7 +62,49 @@ def get_training_metrics():
...
+    """Metric with support for both 1.13 and 1.14+"""
+
+    def reset_states(self):
+        return tf.keras.backend.batch_set_value([(v, 0) for v in self.variables])
+
+    def add_weight(
+        self,
+        name,
+        shape=(),
+        aggregation=tf.VariableAggregation.NONE,
```
Comment (lgeiger) [inline larq/metrics.py]: I'd prefer not changing the default here, so that `LarqMetric` will behave the same as `keras.metrics.Metric`.
```suggestion
        aggregation=tf.VariableAggregation.SUM,
```

File: larq/metrics.py (hunk 8 above)
Comment (lgeiger) [reply larq/metrics.py]: I think switching to `ONLY_FIRST_REPLICA` makes sense here :+1:

File: larq/metrics.py [hunk 10]
```diff
@@ -62,7 +62,49 @@ def get_training_metrics():
     return _GLOBAL_TRAINING_METRICS
 
 
-class FlipRatio(Metric):
+class LarqMetric(Metric):
+    """Metric with support for both 1.13 and 1.14+"""
+
+    def reset_states(self):
```
Comment (lgeiger) [inline larq/metrics.py]: We can also remove this method since it is implemented in the parent class already:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L203-L209

Comment (jneeven) [reply larq/metrics.py]: Agreed, had not noticed this
